core/services/ocr2/plugins/s4/plugin.go

```diff
@@ -0,0 +1,288 @@
+	if config.NSnapshotShards == 0 {
+		return nil, errors.New("number of snapshots shards cannot be zero")
+	}
+	if config.MaxObservationEntries == 0 {
+		return nil, errors.New("max number of observation entries cannot be zero")
+	}
+	if config.MaxReportEntries == 0 {
+		return nil, errors.New("max number of report entries cannot be zero")
+	}
+
+	addressRange := new(big.Int).Sub(s4.MaxAddress, s4.MinAddress)
+	divisor := big.NewInt(int64(config.NSnapshotShards))
+	addressInteval := new(big.Int).Div(addressRange, divisor)
+
+	return &plugin{
+		config:          config,
+		orm:             orm,
+		minQueryAddress: s4.MinAddress,
+		maxQueryAddress: new(big.Int).Add(s4.MinAddress, addressInteval),
+		addressInteval:  addressInteval,
+	}, nil
+}
+
+func (c *plugin) advanceAddressRange() {
+	if c.config.NSnapshotShards > 1 {
+		c.minQueryAddress = new(big.Int).Add(c.minQueryAddress, c.addressInteval)
+		c.maxQueryAddress = new(big.Int).Add(c.maxQueryAddress, c.addressInteval)
+		if c.maxQueryAddress.Cmp(s4.MaxAddress) > 0 {
+			c.maxQueryAddress = s4.MaxAddress
+		}
+		if c.minQueryAddress.Cmp(s4.MaxAddress) >= 0 {
+			c.minQueryAddress = s4.MinAddress
+			c.maxQueryAddress = new(big.Int).Add(c.minQueryAddress, c.addressInteval)
+		}
+	}
+}
+
+func (c *plugin) Query(ctx context.Context, _ types.ReportTimestamp) (types.Query, error) {
+	snapshot, err := c.orm.GetSnapshot(c.minQueryAddress, c.maxQueryAddress, pg.WithParentCtx(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	if len(snapshot) == 0 {
+		return nil, nil
+	}
+
+	rows := make([]*Row, len(snapshot))
+	for i, row := range snapshot {
+		rows[i] = convertRow(row)
+	}
+	query, err := marshalRows(rows, c.minQueryAddress, c.maxQueryAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	c.advanceAddressRange()
+
+	return query, err
+}
+
+func (c *plugin) Observation(ctx context.Context, _ types.ReportTimestamp, query types.Query) (types.Observation, error) {
+	if err := c.orm.DeleteExpired(pg.WithParentCtx(ctx)); err != nil {
+		return nil, err
+	}
+
+	queryRows, minAddress, maxAddress, err := unmarshalRows(query)
+	if err != nil {
+		return nil, err
+	}
+
+	snapshot, err := c.orm.GetSnapshot(minAddress, maxAddress, pg.WithParentCtx(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	snapshotMap := make(map[key]*Row)
+	unconfirmedMap := make(map[key]*Row)
+	for _, row := range snapshot {
+		r := convertRow(row)
+		mkey := key{
+			address: r.Address,
+			slot:    uint(r.Slotid),
+		}
+		snapshotMap[mkey] = r
+		if !row.Confirmed {
+			unconfirmedMap[mkey] = r
+		}
+	}
+
+	observation := make([]*Row, 0)
+	for _, queryRow := range queryRows {
+		mkey := key{
+			address: queryRow.Address,
+			slot:    uint(queryRow.Slotid),
+		}
+		snapshotRow, ok := snapshotMap[mkey]
+		if ok && queryRow.Version < snapshotRow.Version {
+			observation = append(observation, queryRow)
+			delete(unconfirmedMap, mkey)
+		}
+	}
+
+	for _, unconfirmed := range unconfirmedMap {
```

Comment:
The plan was to send out all NEW entires, even from outside the snapshot. Snapshot is only a way of requesting updates and should be independent of propagating NEW entries.

Reply:
That's the plan. We only exclude NEW for two criteria:
- if we see a higher confirmed version,
- the row fits the current round address range (since are building scalable solution).

Reply:
You will also exclude them if you exceed the message size limit.
I don't think NEW entires should be affected by the address range because a malicious leader can choose a specific range on purpose to exclude some entries.

Reply:
Always picking all NEW rows does not scale (we may have too many of them). That's why we need to apply address range here. The entire OCR round should be consistent and operate within the same address range. IMO.

Reply:
BTW, what if we use `types.ReportTimestamp`'s `Epoch` or `Round` to calculate an address range? This would prevent leader from manipulating the range.